### PR TITLE
Fix issues in discovering Test Agent

### PR DESCRIPTION
**Pull request by nigurr**
* Fixed issues in discovering the test agent registry lookup
* added common ps which will be used across DTA flow to discover test agent config
* will work with any version of test agent 14.0/15.0

### CODE REVIEW

File: Tasks/DeployVisualStudioTestAgent/CheckTestAgentInstallation.ps1 [hunk 1]
```diff
@@ -1,93 +1,16 @@
...
-			}
-		}
-	}
-
-	return $false
+    $agentPath = Locate-TestAgent
+    if($agentPath)
+    {
+        $isProductExists = (Get-ChildItem $agentPath.Path).GetValue('Install') -eq '1'
+        if($isProductExists)
```
Comment (allendm-msft) [inline Tasks/DeployVisualStudioTestAgent/CheckTestAgentInstallation.ps1]: remove the blank line

File: Tasks/DeployVisualStudioTestAgent/DeployTestAgent.ps1 [hunk 2]
```diff
@@ -88,6 +96,7 @@ if (!$personalAccessToken)
 try
 {
     $persistExists = CmdletHasMember "Persist"
+    $locateTestAgentHelperExists = DtaCmdletHasMember "LocateTestAgentHelperScriptLocation"
```
Comment (allendm-msft) [inline Tasks/DeployVisualStudioTestAgent/DeployTestAgent.ps1]: locateTestAgentHelperParamExists

File: Tasks/DeployVisualStudioTestAgent/LocateTestAgentHelper.ps1 [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    Write-Verbose "Locating Test Agent: $ProductVersion" -verbose
+    $testAgentDetails = @{}
+    $version = $null
+    $path = $null
+
+    #Find the latest version
+    $regPath = "HKLM:\SOFTWARE\Microsoft\DevDiv\vstf\Servicing"
+    if (-not (Test-Path $regPath))
+    {
+        $regPath = "HKLM:\SOFTWARE\Wow6432Node\Microsoft\DevDiv\vstf\Servicing"
```
Comment (allendm-msft) [inline Tasks/DeployVisualStudioTestAgent/LocateTestAgentHelper.ps1]: wont this fail on a 32bit OS/machine?

File: Tasks/DeployVisualStudioTestAgent/LocateTestAgentHelper.ps1 [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    }
+    if (Test-Path $regPath)
+    {
+        $keys = Get-ChildItem $regPath | Where-Object {$_.GetSubKeyNames() -contains "testagentcore"}
+        $version = Get-SubKeysInFloatFormat $keys | Sort-Object -Descending | Select-Object -First 1
+        $path = $regPath + "\" + $version + "\" + "testagentcore"
```
Comment (allendm-msft) [inline Tasks/DeployVisualStudioTestAgent/LocateTestAgentHelper.ps1]: shouldnt we be using join-path here too?
registry key separators can also change per locale

File: Tasks/DeployVisualStudioTestAgent/LocateTestAgentHelper.ps1 [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        $targetKeys += [decimal] $key.PSChildName
+    }
+
+    return $targetKeys
+}
+
+function Locate-TestAgentPath($ProductVersion = "14.0")
+{
+    $avlVersion = Locate-TestVersion
+    if($avlVersion)
```
Comment (allendm-msft) [inline Tasks/DeployVisualStudioTestAgent/LocateTestAgentHelper.ps1]: so if we dont get it from either registry key shouldnt we fail?

File: Tasks/DeployVisualStudioTestAgent/LocateTestAgentHelper.ps1 [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    $testAgentPath = "HKLM:\SOFTWARE\Wow6432Node\Microsoft\VisualStudio\{0}\EnterpriseTools\QualityTools" -f $ProductVersion
+    if ((Test-Path $testAgentPath) -and (Get-Item $testAgentPath).GetValue('InstallDir'))
+    {
+        return $testAgentPath;
+    }
+
+    Write-Verbose "Test Agent doesn't exist as Path doesn't exist" -verbose
+    return $testAgentPath;
```
Comment (allendm-msft) [inline Tasks/DeployVisualStudioTestAgent/LocateTestAgentHelper.ps1]: tests for this new ps1 and other changes?

**Comment by allendm-msft** [comment]
lets add the tests

File: Tasks/DeployVisualStudioTestAgent/CheckTestAgentInstallation.ps1 (hunk 1 above)
Comment (nigurr) [inline Tasks/DeployVisualStudioTestAgent/CheckTestAgentInstallation.ps1]: I didn't notice any blank line

File: Tasks/DeployVisualStudioTestAgent/LocateTestAgentHelper.ps1 (hunk 5 above)
Comment (nigurr) [inline Tasks/DeployVisualStudioTestAgent/LocateTestAgentHelper.ps1]: We do return null and the caller is handling error.

File: Tasks/DeployVisualStudioTestAgent/LocateTestAgentHelper.ps1 (hunk 4 above)
Comment (nigurr) [inline Tasks/DeployVisualStudioTestAgent/LocateTestAgentHelper.ps1]: Registry Entries are always separated by "\" irrespective of locale/language

**Comment by allendm-msft** [comment]
Approved

**Comment by allendm-msft** [comment]
Approved with comments

**Pull request closed without merging**